src/DataStructures/Tensor/Tensor.hpp

```diff
@@ -589,137 +589,110 @@ std::ostream& operator<<(
-  apply(const Tensor<DataVector, Structure...>& input,
-        const double value) noexcept {
-    return DataVector(input.begin()->size(), value);
+template <typename T, typename... StructureOut, typename U,
+          typename... StructureIn>
+struct MakeWithValueImpl<Tensor<T, StructureOut...>, Tensor<U, StructureIn...>,
+                         Requires<Tensor_detail::is_scalar<T>>> {
+  /// \brief Returns a Tensor whose elements are set equal to `value` (`input`
+  /// is ignored).
+  static SPECTRE_ALWAYS_INLINE Tensor<T, StructureOut...> apply(
+      const Tensor<U, StructureIn...>& /*input*/, const T value) noexcept {
+    return Tensor<T, StructureOut...>(value);
   }
 };
 
-template <typename... StructureOut, typename... StructureIn>
-struct MakeWithValueImpl<Tensor<DataVector, StructureOut...>,
-                         Tensor<DataVector, StructureIn...>> {
+template <typename T, typename... Structure, typename U>
+struct MakeWithValueImpl<Tensor<T, Structure...>, U,
+                         Requires<(Tensor_detail::is_vector<T> or
+                                   Tensor_detail::is_spin_weighted<T>) and
+                                  Tensor_detail::is_vector<U>>> {
+  /// \brief Returns a Tensor whose DataVectors are the same size as `input`,
+  /// with each element equal to `value`.
+  static SPECTRE_ALWAYS_INLINE Tensor<T, Structure...> apply(
+      const U& input, const double value) noexcept {
+    return Tensor<T, Structure...>(input.size(), value);
+  }
+};
+
+template <typename T, typename... StructureOut, typename U,
+          typename... StructureIn>
+struct MakeWithValueImpl<Tensor<T, StructureOut...>, Tensor<U, StructureIn...>,
+                         Requires<(Tensor_detail::is_vector<T> or
+                                   Tensor_detail::is_spin_weighted<T>) and
+                                  Tensor_detail::is_vector<U>>> {
   /// \brief Returns a Tensor whose DataVectors are the same size as the
   /// DataVectors of `input`, with each element equal to `value`.
-  static SPECTRE_ALWAYS_INLINE Tensor<DataVector, StructureOut...> apply(
-      const Tensor<DataVector, StructureIn...>& input,
-      const double value) noexcept {
-    return Tensor<DataVector, StructureOut...>(input.begin()->size(), value);
+  static SPECTRE_ALWAYS_INLINE Tensor<T, StructureOut...> apply(
+      const Tensor<U, StructureIn...>& input, const double value) noexcept {
+    return Tensor<T, StructureOut...>(input.begin()->size(), value);
   }
 };
 
-template <typename... Structure>
-struct MakeWithValueImpl<Tensor<double, Structure...>, double> {
-  /// \brief Returns a Tensor whose elements are set equal to `value` (`input`
-  /// is ignored).
-  static SPECTRE_ALWAYS_INLINE Tensor<double, Structure...> apply(
-      const double /*input*/, const double value) noexcept {
-    return Tensor<double, Structure...>(value);
+// Creating a plain double or complex<double> is handled by generic versions.
```

Comment:
Could you point to where this generic version is implemented? I would have expected the overload that creates a `double` from a `Tensor<double>` would be located in `Tensor.hpp` (so having this comment is useful).